### PR TITLE
Implemented sRGB support for DXT encoded textures

### PR DESCRIPTION
**Pull request by BStringhamVRSK**
Fixes [issue 1497](https://github.com/google/filament/issues/1497)

This change allows the user to specify gamma encoded (sRGB) DXT textures.
Changes have been made for OpenGL, Vulkan, and Metal, but only the OpenGL backend has been tested.  Given the nature of the changes, I don't expect the other backends to have problems.

### CODE REVIEW

**Comment by romainguy** [comment]
Thank you for your contribution! Your change breaks the WebGL build unfortunately:

```
In file included from ../../filament/backend/src/opengl/OpenGLContext.h:24:
../../filament/backend/src/opengl/GLUtils.h:381:55: error: use of undeclared identifier 'GL_COMPRESSED_SRGB_S3TC_DXT1_EXT'
        case TextureFormat::DXT1_SRGB:         return GL_COMPRESSED_SRGB_S3TC_DXT1_EXT;
                                                      ^
../../filament/backend/src/opengl/GLUtils.h:383:55: error: use of undeclared identifier 'GL_COMPRESSED_SRGB_ALPHA_S3TC_DXT1_EXT'
        case TextureFormat::DXT1_SRGBA:        return GL_COMPRESSED_SRGB_ALPHA_S3TC_DXT1_EXT;
                                                      ^
../../filament/backend/src/opengl/GLUtils.h:385:55: error: use of undeclared identifier 'GL_COMPRESSED_SRGB_ALPHA_S3TC_DXT3_EXT'
        case TextureFormat::DXT3_SRGBA:        return GL_COMPRESSED_SRGB_ALPHA_S3TC_DXT3_EXT;
                                                      ^
../../filament/backend/src/opengl/GLUtils.h:387:55: error: use of undeclared identifier 'GL_COMPRESSED_SRGB_ALPHA_S3TC_DXT5_EXT'
        case TextureFormat::DXT5_SRGBA:        return GL_COMPRESSED_SRGB_ALPHA_S3TC_DXT5_EXT;
```
For WebGL 2.0 and OpenGL ES 3.0 you need to test whether the extension (EXT_texture_compression_s3tc_srgb)[https://www.khronos.org/registry/OpenGL/extensions/EXT/EXT_texture_compression_s3tc_srgb.txt] is present.

**Comment by BStringhamVRSK** [comment]
I’m looking into a fix.  I’m trying to figure out why those values aren’t defined in the WebGL build.  I think it might be an include file ordering thing?  Or, something needs a #define.  We’ll see.  I’m hoping to get it resolved soon.

-Brian

**Comment by BStringhamVRSK** [comment]
Oh, I shouldn’t have responded before I finished reading your message!  Thanks for the heads-up.

From: Stringham, Brian
Sent: Monday, January 13, 2020 5:06 PM
To: google/filament <reply@reply.github.com>
Subject: RE: [google/filament] Implemented sRGB support for DXT encoded textures (#2026)

I’m looking into a fix.  I’m trying to figure out why those values aren’t defined in the WebGL build.  I think it might be an include file ordering thing?  Or, something needs a #define.  We’ll see.  I’m hoping to get it resolved soon.

-Brian

File: filament/backend/src/opengl/OpenGLDriver.cpp [hunk 1]
```diff
@@ -1286,7 +1286,12 @@ bool OpenGLDriver::isTextureFormatSupported(TextureFormat format) {
         return gl.ext.texture_compression_etc2;
     }
     if (isS3TCCompression(format)) {
-        return gl.ext.texture_compression_s3tc;
+        if (isS3TCSRGBCompression(format))		{
```
Comment (romainguy) [inline filament/backend/src/opengl/OpenGLDriver.cpp]: Formatting issues: extra spaces on this line, `else {` should be on the same line as the closing `}` from the if

File: filament/backend/src/opengl/OpenGLContext.cpp [hunk 2]
```diff
@@ -189,6 +189,7 @@ void OpenGLContext::initExtensionsGLES(GLint major, GLint minor, ExtentionSet co
     ext.texture_compression_s3tc = hasExtension(exts, "WEBGL_compressed_texture_s3tc");
     ext.EXT_multisampled_render_to_texture = hasExtension(exts, "GL_EXT_multisampled_render_to_texture");
     ext.KHR_debug = hasExtension(exts, "GL_KHR_debug");
+    ext.EXT_texture_sRGB = hasExtension(exts, "GL_EXT_texture_sRGB");
```
Comment (romainguy) [inline filament/backend/src/opengl/OpenGLContext.cpp]: `EXT_texture_sRGB` is a not a GL ES extension. The equivalent in GL ES is [`EXT_texture_sRGB_decode`](https://www.khronos.org/registry/OpenGL/extensions/EXT/EXT_texture_sRGB_decode.txt).

**Comment by romainguy** [comment]
Adding @bejado and @prideout to check the Vulkan and Metal implementations.

File: filament/backend/src/opengl/OpenGLContext.cpp (hunk 2 above)
Comment (BStringhamVRSK) [reply filament/backend/src/opengl/OpenGLContext.cpp]: Thanks for catching this.  I think the one we want is: GL_EXT_texture_compression_s3tc_srgb, which I have now added.

File: filament/backend/src/opengl/OpenGLDriver.cpp (hunk 1 above)
Comment (BStringhamVRSK) [reply filament/backend/src/opengl/OpenGLDriver.cpp]: Thanks.  I've pushed up this change.

**Comment by romainguy** [comment]
Thank you for your contribution!

**Pull request merged**